### DISCUSSION TITLE
Is it possible to have commands with generic state?

### DISCUSSION THREAD

**Discussion by ls-ivan-kiselev** [tauri-apps/tauri#7444] [Q&A]
So, my initial problem is unit-testing of tauri commands that do some IO, such as API calls in various directions that are rather hard targets for real integration testing.

I was trying to achieve it with generic manage state, minimal code:

```Rust
use tauri::State;

trait ApiClientLike: Sync + Send {}

struct RealApiClient{}
impl ApiClientLike for RealApiClient{}

#[tauri::command]
fn api_call<T: ApiClientLike>(state: State<'_, T>) {
    todo!("Here we do some API calls that are defined on the trait")
}

fn main() {
    tauri::Builder::default()
        .invoke_handler(tauri::generate_handler![api_call])
        .manage(RealApiClient{})
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
}


#[cfg(test)]
mod tests {
    use super::*;
    struct MockApiClient{}
    impl ApiClientLike for MockApiClient {}

    // Some tests with tauri::test, but MyMockState as .manage(MockApiClient{})
}
```

But it turns out into the error:
```Rust
   Compiling demo-generics-in-commands v0.0.0 (REDACTED)
error[E0282]: type annotations needed
  --> src/main.rs:8:1
   |
8  | #[tauri::command]
   | ^^^^^^^^^^^^^^^^^ cannot infer type
...
15 |         .invoke_handler(tauri::generate_handler![api_call])
   |                         ---------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `__cmd__api_call` which comes from the expansion of the macro `tauri::generate_handler` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0282`.
error: could not compile `demo-generics-in-commands` (bin "demo-generics-in-commands") due to previous error
```

And, well, I am not sure how go around this problem, thus seeking an advice. 

I completely have failed to find something similar in discussions/issues, which leads me to a suspicion that people either don't unit-test tauri commands that perform IO, or just simply implement them in a completely different way, which puzzles me even more.

**Comment by FabianLars** [comment]
> Is it possible to have commands with generic state?

afaik, no. There's no way for Rust to figure out the actual types here (which it has to).

> which leads me to a suspicion that people either don't unit-test tauri commands that perform IO

I doubt that, but from what i've seen unit testing (on the rust side) is indeed kinda nieche still :/ Also, i'm not sure what kind of IO operations you're exactly talking about, but did you also play around with good old type aliases like
```rs
#[cfg(not(test))]
type Client = RealApiClient;
#[cfg(test)]
type Client = TestApiClient;
```
or alternatively use the same client struct but have 2 impl blocks which you toggle on or off with the same cfg flag.

-> I didn't do much unit testing in my own apps (mainly because i didn't really touch them in the last 2 years...) so that may show in my response 😅

**Reply by ls-ivan-kiselev** [reply]
> Also, I'm not sure what kind of IO operations you're exactly talking about

In my case it's networking calls to Github API and AWS API, which I can mock and isolate in unit-tests, but having testing infrastructure mocking Github and AWS over the network to point my rust code to, with purpose of testing, is just expensive in all the senses. 

> but did you also play around with good old type aliases

No I haven't! I didn't even think that you can pull such a trick in Rust! I will definitely give it a try, thanks a lot! 
To be reported later in the thread.
